forc-test/src/lib.rs

```diff
@@ -353,16 +451,41 @@ impl BuiltTests {
 /// First builds the package or workspace, ready for execution.
 pub fn build(opts: Opts) -> anyhow::Result<BuiltTests> {
     let build_opts = opts.into_build_opts();
+    let build_plan = pkg::BuildPlan::from_build_opts(&build_opts)?;
     let built = pkg::build_with_options(build_opts)?;
-    BuiltTests::from_built(built)
+    let mut built_members = built.clone().into_members()?;
+
+    // For each member node collect their contract dependencies.
+    let member_contract_dependencies: HashMap<pkg::Pinned, Vec<pkg::BuiltPackage>> = build_plan
+        .member_nodes()
+        .map(|member_node| {
+            let graph = build_plan.graph();
+            let pinned_member = graph[member_node].clone();
+            let pinned_contract_dependencies: Vec<pkg::Pinned> = build_plan
+                .contract_dependencies(member_node)
+                .map(|contract_depency_node_ix| graph[contract_depency_node_ix].clone())
+                .collect();
+            let contract_dependencies = pinned_contract_dependencies
+                .iter()
+                .filter_map(|pinned| built_members.remove(pinned))
+                .collect();
+
+            (pinned_member, contract_dependencies)
+        })
+        .collect();
```

Comment:
It looks like the produced `member_contract_dependencies` might not always contain all dependencies in the case that a member contract is depended on by two other contracts, as it will be removed from the `built_members` set the first time it appears in a member's `pinned_contract_dependencies`?

Also, it looks like this `clone`s all the built members at the start, which is quite a heavy operation (i.e. all bytecode, etc). Ideally we shouldn't have to clone built packages at all, and can reference the original members via keys or similar.

It looks like we might benefit from a refactor between this `build` constructor and the `BuiltTests::from_built` constructor to achieve what we want much more efficiently (and without missing contract deps for some members).

Reply:
Thanks for pointing it, I wasn't happy how this turned out but couldn't found a way to avoid clones just by refactoring `forc-test`. To solve it I made a small refactor in `pkg::Built`.

While trying to avoid cloning I realized the way we store `Built::Package` (which was a `Box<BuiltPackage>`) and `Built::Workspace` (which was a `HashMap<String, BuiltPackage>`) made this very hard to deal with.  Also storing both the name and built package felt wrong as built package already contains its pinned counterpart.

In a870681e43a48869e58f82a32b252b27fa12e33f i refactored them to hold `Arc<_>` and `Vec<Arc<_>>` and by doing that I am avoiding entire clones of `BuiltPackage` struct. Also refactored `inner_member()` just return an iterator so that the caller can decide if they want to collect it or not. 

After this change I tested with my simple workspace test containing couple of contracts and although I wasn't expecting it to be the effect was visible 😄 it is about ~%10 faster in the total time whole test operation takes. I think the effect is visible since std and core is considerably big so cloning them was taking some time.